File: core/flows/src/flows_steps/branding.py
"""branding.py — WHAT THIS DEPLOYMENT CALLS ITSELF, for the mail this engine sends.

The same `branding` platform setting the terminal reads, read here for the two names a recipient
actually sees: the From display name on every message, and the product's name where a sentence
has to say it. One store, so the workbench header and the mail signature cannot disagree about
what this product is called — which is precisely what happened while the name was a literal in
`emailx.send` and a build arg in the terminal image.

READ THROUGH IDENTITY, like every other deployment fact this tier needs: `admin-api` owns
platform settings and is the one domain everybody may depend on. Cached for a short window
because mail is bursty — a twenty-person fan-out is twenty sends in a second and they must not be
twenty settings reads — and because an operator who renames the product expects the NEXT meeting's
mail to carry it, not this one's second half.

FAILS SOFT, ALWAYS. admin-api down, the key unset, a half-written value: every one of them
answers the stock names. A deployment that cannot reach its settings store still has minutes to
deliver, and a mail that does not send because the sender's DISPLAY NAME could not be looked up is
the worst possible trade.
"""
from __future__ import annotations

import time
from typing import Optional

# NOT `from .common import ADMIN_API`. That name is a REQUIRED-EXPLICIT door served by `common`'s
# PEP-562 `__getattr__`, so importing it resolves the door AT IMPORT — the thing `_door` exists to
# avoid, and the thing `flows_defs/production.py` has a paragraph about. Resolved per call below.
from .common import _door, http, require_admin_key

#: The stock names — the product's own, used when nothing is configured.
DEFAULT_PRODUCT = "Vexa"

TTL_S = 60.0
_CACHE: Optional[tuple[dict, float]] = None


def reset_cache() -> None:
    """Forget the cached answer. TEST seam, and the same shape `instance_gate.reset_cache` has."""
    global _CACHE
    _CACHE = None


def _read() -> dict:
    """One live read of the `branding` key. Never raises — every failure IS the empty answer."""
    try:
        door = _door("VEXA_FLOWS_ADMIN_API_URL").rstrip("/")
        code, body = http("GET", f"{door}/internal/settings/branding",
                          {"X-Internal-Secret": _internal_secret()}, timeout=5)
    except Exception:  # noqa: BLE001 — including StepError, which common.http raises
        return {}
    if code != 200 or not isinstance(body, dict):
        return {}
    value = body.get("value")
    return value if isinstance(value, dict) else {}


def _internal_secret() -> str:
    # Imported lazily and by name so a deployment that has not set it fails HERE, with the empty
    # answer above, rather than at import of this module — branding is never worth a boot failure.
    from .common import require_internal_secret
    try:
        return require_internal_secret()
    except Exception:  # noqa: BLE001
        return ""


def settings() -> dict:
    """The stored branding dict, cached for `TTL_S`. `{}` when there is none."""
    global _CACHE
    now = time.monotonic()
    if _CACHE is not None and now < _CACHE[1]:
        return _CACHE[0]
    value = _read()
    _CACHE = (value, now + TTL_S)
    return value


def product_name() -> str:
    """What this deployment is called. Never empty."""
    return str(settings().get("product_name") or "").strip() or DEFAULT_PRODUCT


def sender_name() -> str:
    """The From DISPLAY name on outbound mail — the explicit setting, else the product's name.

    An operator who renamed the product and nothing else means the mail to come from the product
    they named; making them type it twice is how the header and the signature drift apart."""
    value = settings()
    return (str(value.get("sender_name") or "").strip()
            or str(value.get("product_name") or "").strip()
            or DEFAULT_PRODUCT)


def bot_name() -> str:
    """The bot's display name in the meeting — the explicit setting, else the product's name."""
    value = settings()
    return (str(value.get("bot_name") or "").strip()
            or str(value.get("product_name") or "").strip()
            or DEFAULT_PRODUCT)


def support_email() -> str:
    """Where a recipient can reach a human, or "" when the deployment names nobody."""
    return str(settings().get("support_email") or "").strip()


def powered_by() -> bool:
    """Whether outbound mail carries the "Built on Vexa" attribution line.

    ON unless an operator explicitly turned it off: the attribution is the licence-friendly
    default, and a typo in the setting must never be the thing that removes it."""
    return str(settings().get("powered_by") or "").strip().lower() not in ("false", "0", "no", "off")


def require_admin_key_present() -> bool:
    """Whether this tier can even ask. Used by nothing yet; kept next to the reader so a future
    caller does not reinvent the check against a different variable."""
    try:
        return bool(require_admin_key())
    except Exception:  # noqa: BLE001
        return False

File: clients/terminal/src/branding/branding.ts
/** branding.ts — WHAT THIS DEPLOYMENT IS CALLED, as one shape both tiers agree on.
 *
 *  The values live in admin-api's `branding` platform setting and are read at RUNTIME, so an
 *  operator renames a deployment with a settings write and no rebuild. That is the whole point:
 *  the name used to be a literal in a dozen components and `NEXT_PUBLIC_DEFAULT_BOT_NAME` was a
 *  BUILD arg, so "call it something else" cost an image.
 *
 *  Server and client both import THIS module — the route below serves `Branding`, the hook
 *  consumes it, and `DEFAULTS` is the one place the stock names are written. Two copies of a
 *  default is how a sign-in page and a header come to disagree about what the product is called.
 */
export interface Branding {
  /** Replaces "Vexa" wherever a person reads the product's name. */
  productName: string;
  /** Hex, validated server-side (admin-api `_HEX_COLOR`) before it ever reaches a stylesheet. */
  accent: string;
  /** An http(s) URL, or "" to fall back to the bundled mark. */
  logoUrl: string;
  /** The bot's display name in the meeting's participant list. */
  botName: string;
  /** The From display name on outbound mail. */
  senderName: string;
  /** Shown in the mail footer; "" hides the line. */
  supportEmail: string;
  /** Whether the "Built on Vexa" attribution line shows. */
  poweredBy: boolean;
}

export const DEFAULTS: Branding = {
  productName: "Vexa",
  accent: "",            // "" = keep the stylesheet's own --accent
  logoUrl: "",
  botName: "Vexa",
  senderName: "Vexa",
  supportEmail: "",
  poweredBy: true,
};

/** admin-api stores strings; this is the one place they become a typed `Branding`.
 *  Unset fields fall back to `DEFAULTS` field-by-field, never all-or-nothing: an operator who
 *  set only the product name still gets every other default rather than an empty product. */
export function fromSetting(value: Record<string, unknown> | null | undefined): Branding {
  const s = (k: string): string => {
    const v = value?.[k];
    return typeof v === "string" ? v.trim() : "";
  };
  const poweredBy = s("powered_by").toLowerCase();
  return {
    productName: s("product_name") || DEFAULTS.productName,
    accent: s("accent") || DEFAULTS.accent,
    logoUrl: s("logo_url") || DEFAULTS.logoUrl,
    // The bot and the sender default to the PRODUCT name rather than to "Vexa": an operator who
    // names the product and nothing else expects the bot in the meeting to introduce the same
    // product, not the one they renamed away from.
    botName: s("bot_name") || s("product_name") || DEFAULTS.botName,
    senderName: s("sender_name") || s("product_name") || DEFAULTS.senderName,
    supportEmail: s("support_email") || DEFAULTS.supportEmail,
    // Only an explicit false hides it; anything else (unset, a typo) shows the line, because the
    // attribution is the licence-friendly default and a typo must not silently remove it.
    poweredBy: !["false", "0", "no", "off"].includes(poweredBy),
  };
}

/** The initial used when no logo is configured — the same glyph the mark would carry. */
export function initial(name: string): string {
  return (name.trim().charAt(0) || "V").toUpperCase();
}
